pkg/controller/v1beta1/inferenceservice/components/predictor.go

```diff
@@ -521,21 +549,101 @@ func multiNodeProcess(sRuntime v1alpha1.ServingRuntimeSpec, isvc *v1beta1.Infere
 	}
 	return mergedWorkerPodSpec, nil
 }
 
+// The `rayNodeCount` is determined based on the requested GPU count.
+// We use the GPU resource defined in `workerSpec` to calculate the required GPU count.
+// The `rayNodeCount` is set to the ceiling value of (total requested GPU count / GPUs per worker node).
+// The head node GPU count is determined by subtracting (workerSpec GPU resource * (rayNodeCount - 1)) from the total GPU count.
+// If the head node has a predefined GPU count, it takes precedence.
+// However, if the total required GPU count is less than the sum of (head node GPU count + workerSpec GPU resource * rayNodeCount), an error is raised.
+func computeRayNodeAndGPUs(mergedWorkerPodSpec *corev1.PodSpec, totalRequestGPUCount int, podSpec *corev1.PodSpec) (int, int, int, error) {
+	getGPUResourceQty := func(spec *corev1.PodSpec) int {
+		if _, gpuResourceQuantity, exist := utils.GetGPUResourceQtyByType(&spec.Containers[0].Resources, "Request"); exist {
+			return int(gpuResourceQuantity.Value())
+		}
+		return 0
+	}
+
+	computeRayNodes := func(totalGPUs, workerGPUs int) int {
+		if workerGPUs == 0 {
+			return 1
+		}
+		return int(math.Ceil(float64(totalGPUs) / float64(workerGPUs)))
+	}
+
+	validateGPUAllocation := func(rayNodeCount, headNodeGPUCount, workerNodeGPUCount int) error {
+		if totalRequestGPUCount > headNodeGPUCount+(workerNodeGPUCount*(rayNodeCount-1)) {
+			return fmt.Errorf(ErrRayClusterInsufficientGPUs, totalRequestGPUCount, headNodeGPUCount, workerNodeGPUCount*(rayNodeCount-1))
+		}
+		return nil
+	}
+
+	var rayNodeCount int
+	headNodeGPUCount := getGPUResourceQty(podSpec)
+	workerNodeGPUCount := getGPUResourceQty(mergedWorkerPodSpec)
+
+	// Case 1 & 2: At least worker GPUs are set
+	if workerNodeGPUCount > 0 {
+		rayNodeCount = computeRayNodes(totalRequestGPUCount, workerNodeGPUCount)
+		newHeadNodeGPUCount := totalRequestGPUCount - workerNodeGPUCount*(rayNodeCount-1)
+
+		if headNodeGPUCount > 0 {
+			rayNodeCountByHeadGpuCount := computeRayNodes(totalRequestGPUCount, headNodeGPUCount)
+			if rayNodeCountByHeadGpuCount == 1 {
+				rayNodeCount = 1
+			}
+		} else if headNodeGPUCount == 0 || headNodeGPUCount == newHeadNodeGPUCount {
+			headNodeGPUCount = newHeadNodeGPUCount
+		}
+
+		// Use only head node with total request gpu count if it can satisfy the GPU requirement
+		if rayNodeCount == 1 {
+			workerNodeGPUCount = 0
+			headNodeGPUCount = totalRequestGPUCount
+		}
+
+		if err := validateGPUAllocation(rayNodeCount, headNodeGPUCount, workerNodeGPUCount); err != nil {
+			return 0, 0, 0, err
+		}
+		return rayNodeCount, workerNodeGPUCount, headNodeGPUCount, nil
+	}
+
+	// Case 3: Only head GPU is set
+	if headNodeGPUCount > 0 {
+		remainingGPUs := totalRequestGPUCount - headNodeGPUCount
+		rayNodeCount = 1
+		workerNodeGPUCount = 1
+
+		if remainingGPUs > 0 {
+			rayNodeCount = computeRayNodes(remainingGPUs, 1) + 1 // Single GPU worker nodes
```

Comment:
Simplification suggestion:

```suggestion
			rayNodeCount += remainingGPUs // Single GPU worker nodes
```

Reply:
This is the same reason why I want to keep original code but I want to update like this
```
rayNodeCount = computeRayNodes(remainingGPUs, workerNodeGPUCount) + 1 // Single GPU worker nodes
```
want to keep the same format that is used in the func